### PR TITLE
fix: build the JOB URL on the server side

### PR DESCRIPTION
**Pull request by mkanoor**
With the advent of controller behind our server the URL to access the job details for JobTemplate and WorkflowTemplate has changed

We have an api end point that is specificed in the RH AAP Credential Type this is sent to the ansible-rulebook so that it can talk to the controller via the api. This end point could point to a controller directly or to a controller behind our server.

To directly access the controller we only need the host and port so the API end point would look like
e.g. https://<<your_controller>>/

This will map the Job URL for a Job Template to look like
     https://<<your_controller>>/#/jobs/playbook/23/details/
This will map the Job URL for a Workflow Template to look like
     https://<<your_controller>>/#/jobs/workflow/24/details/

When the controller is behind our server the API end point is e.g. https://<<your_server>>/api/controller

This will map the Job URL for a Job Template to look like
     https://<<your_server>>/execution/jobs/playbook/25/details/
This will map the Job URL for a Workflow Template to look like
     https://<<your_server>>/execution/jobs/workflow/26/details/

For this to work correctly we need ansible-rulebook to pass in the controller_job_id and the action_type. Currently it doesn't send up the controller_job_id which is being fixed in a separate [PR](https://github.com/ansible/ansible-rulebook/pull/698).

https://issues.redhat.com/browse/AAP-25604

### CODE REVIEW

File: src/aap_eda/wsapi/messages.py [hunk 1]
```diff
@@ -46,6 +46,7 @@ class ActionMessage(Message):
     delay: Optional[float]
     message: Optional[str]
     kind: Optional[str]
+    controller_job_id: Optional[str]
```
Comment (hsong-rh) [inline src/aap_eda/wsapi/messages.py:49]: What's the difference between this new `controller_job_id` and the existing `job_id`?

File: src/aap_eda/wsapi/consumers.py [hunk 2]
```diff
@@ -417,3 +435,37 @@ def get_eda_system_vault_passwords(
...
         return vault_passwords
+
+    def _get_url(self, message: ActionMessage, inputs: dict) -> str:
+        if message.action not in ("run_job_template", "run_workflow_template"):
+            return ""
+        url = message.url
+        if not message.controller_job_id:
+            return url
+
+        api_url = inputs["host"]
```
Comment (Alex-Izquierdo) [inline src/aap_eda/wsapi/consumers.py:446]: It seems that this method can be called with empty `inputs`. Should not be handled the potential absence of the "host" key?

File: tests/integration/wsapi/test_consumer.py [hunk 3]
```diff
@@ -448,6 +448,119 @@ async def test_multiple_rules_for_one_event(
         assert await get_audit_event_action_count(event) == 2
 
 
+test_data = [
```
Comment (Alex-Izquierdo) [inline tests/integration/wsapi/test_consumer.py:451]: Can you use a more specific name?

File: src/aap_eda/wsapi/messages.py (hunk 1 above)
Comment (mkanoor) [reply src/aap_eda/wsapi/messages.py:49]: @Alex-Izquierdo I thought they are the same but it predates us and is a unique job identifier for every action that we run. Initially I thought it is the job id from controller only to realize its a UUID and not the controller job id

Comment (Alex-Izquierdo) [reply src/aap_eda/wsapi/messages.py:49]: Thanks for the clarification, but the question was made by @hsong-rh xD

File: src/aap_eda/wsapi/consumers.py (hunk 2 above)
Comment (hsong-rh) [reply src/aap_eda/wsapi/consumers.py:446]: For the input of credential type of AAP, the `hosts` is required

Comment (Alex-Izquierdo) [reply src/aap_eda/wsapi/consumers.py:446]: the function can still be called with empty "inputs": https://github.com/ansible/eda-server/pull/948/files#diff-179935aa48d00084c8f2722f9b007cad6d09b24425ffe4121cc75f6c37b03c42R279

File: src/aap_eda/wsapi/consumers.py [hunk 4]
```diff
@@ -417,3 +435,37 @@ def get_eda_system_vault_passwords(
...
+        path = urlparts.path.rstrip("/")
+        if path == "":
+            path = "/"
+        if path in settings.API_PATH_TO_UI_PATH_MAP:
+            path = settings.API_PATH_TO_UI_PATH_MAP[path]
+
+        if message.action == "run_job_template":
+            slug = f"{path}/jobs/playbook/{message.controller_job_id}/details/"
+        elif message.action == "run_workflow_template":
+            slug = f"{path}/jobs/workflow/{message.controller_job_id}/details/"
```
Comment (jshimkus-rh) [inline src/aap_eda/wsapi/consumers.py:458]: Should be `if/else`.  The check at entry guarantees it's one or the other.
As written it gives the impression that `slug` may not be set before use in `urlunparse`.

**Review by Alex-Izquierdo — APPROVED** [review]

**Review by jshimkus-rh — APPROVED** [review]

**Pull request merged**